### ISSUE TITLE
`repeated()` requiring `collect()`

### ISSUE THREAD

**Issue by cengels** [zesterer/chumsky#392]
Hey there! I'm currently working on rewriting my parser to the new `zero-copy` alpha version, but I hit a bit of a stumbling block in my parsers that generate an `IterParser` such as `repeated()` or `separated_by()`.

Specifically, the fact that you now need to explicitly call `collect::<Vec<_>>()` afterwards (or `foldr()`), otherwise the result of the parser is simply `()`.

My first question is therefore this: is there a way to improve the parser's result type in this case? If there's no way to implicitly have `IterParser`s return an iterator when you do things like try to `map()` their result, would it be possible to at least change the return type to something less cryptic, or perhaps use trait bounds to generate an appropriate compiler error when one attempts to process the result of an `IterParser` without first calling `collect()`?

Personally, I'm ashamed to say it took me far too long to understand why my parser was suddenly returning no result. I had to reread the new version of the guide quite a few times until I spotted that every example that uses `repeated()` or `separated_by()` now uses `collect::<Vec<_>>()` directly after it. This solved the problem, but allocating an extra `Vec` when all I really want is an iterator doesn't feel ideal, even if the actual performance cost of doing this is probably completely insignificant.

I did discover the method `IterParser::parse_iter()`, which, looking at the method signature and the brief description, does seem to be exactly what I want, but as it requires an extra argument `input` I don't quite understand how you're supposed to use it, and so far there seems to be no example (either in the documentation or the `examples` folder) that utilizes `parse_iter()`.

To clarify, instead of this:

```
my_parser
    .repeated()
    .collect::<Vec<_>>()
    .map(|v| v.into_iter().map(|x| ...));
```

I'd ideally like to be able to write

```
my_parser
    .repeated()
    .map(|v| v.map(|x| ...));
```

or even use a special method, like so:

```
my_parser
    .repeated()
    .into_iter().map(|v| v.map(|x| ...));
```

Since I have very little understanding of the parser internals, I cannot judge how feasible this is. If it _isn't_ feasible, would it at least be possible to add a sentence or two about this behaviour in the documentation for `repeated()` and `separated_by()`? I fear I won't be the only one stumbling upon this and wondering why their parser is only returning `()`.

**Comment by zesterer** [comment]
Hey, sorry you've been having issues with this

> is there a way to improve the parser's result type in this case? If there's no way to implicitly have IterParsers return an iterator when you do things like try to map() their result

Unfortunately it's not possible to output an iterator because the parser still needs to perform error handling internally.

> would it be possible to at least change the return type to something less cryptic

It has been suggested in the past that `repeated` should output a slice of its input by default. Although I'm somewhat empathetic toward this, it requires that the input implements `SliceInput`, and there's no guarantee of that.

> or perhaps use trait bounds to generate an appropriate compiler error when one attempts to process the result of an IterParser without first calling collect()?

Unfortunately it's extremely difficult to coerce Rust's trait errors into something more sensible. I hope that as chumsky becomes popular, it puts (friendly) pressure on the core Rust devs to find ways to simplify such errors.

An option might be to remove the `Parser` impl for `Repeated`, but this breaks parsers that *don't* care about their output value, like [this](https://github.com/zesterer/chumsky/blob/main/examples/nano_rust.rs#L58-L61).

> This solved the problem, but allocating an extra Vec when all I really want is an iterator doesn't feel ideal

You can map the output before it goes through `repeated`, like so:

```rs
my_parser
    .map(|x| ...)
    .repeated()
```

> If it isn't feasible, would it at least be possible to add a sentence or two about this behaviour in the documentation for repeated() and separated_by()?

Definitely. `1.0` is still in alpha state and minor things (especially documentation) are still being worked on. It'll be doing a pass over the docs to correct any inaccuracies and clarify details like this before release.

**Comment by cengels** [comment]
> You can map the output before it goes through `repeated`, like so:
> 
> ```rust
> my_parser
>     .map(|x| ...)
>     .repeated()
> ```

That is absolutely true. I was hesitant to do this at first because I generally prefer to clearly separate the parsing process and the output mapping, but this seems like the best option for now. Plus, I could still separate the "single item" parser and the corresponding `repeated()` parser into two different definitions, which would also solve the SoC problem in this case.

I noticed you evaded my mention of `IterParser::parser_iter()`. Does that mean this is not a method intended for public use (yet)? I have to admit, even if it's totally unrelated to the initial question, now that I've found it I am quite curious what that method is intended to be used for.

Anyway, thanks for the fast response and all the work you've done on this library!
